src/System.Net.Sockets/tests/FunctionalTests/SocketAsyncEventArgsTest.cs

```diff
@@ -61,5 +61,66 @@ public void TestDefaultConstructor()
                 }
             }
         }
+
+        [OuterLoop] // TODO: Issue #11345
+        [Fact]
+        public async Task ReuseSocketAsyncEventArgs_MutateBufferList()
+        {
+            using (var listen = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listen.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listen.Listen(1);
+
+                Task<Socket> acceptTask = listen.AcceptAsync();
+                await Task.WhenAll(
+                    acceptTask,
+                    client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listen.LocalEndPoint).Port)));
+
+                using (Socket server = await acceptTask)
+                {
+                    TaskCompletionSource<bool> tcs = null;
+
+                    var sendBuffer = new byte[64];
+                    var sendBufferList = new ArraySegment<byte>[1] { new ArraySegment<byte>(sendBuffer, 0, 1) };
+                    var sendArgs = new SocketAsyncEventArgs();
+                    sendArgs.BufferList = sendBufferList;
+                    sendArgs.Completed += (_,__) => tcs.SetResult(true);
+
+                    var recvBuffer = new byte[64];
+                    var recvBufferList = new ArraySegment<byte>[1] { new ArraySegment<byte>(recvBuffer, 0, 1) };
+                    var recvArgs = new SocketAsyncEventArgs();
+                    recvArgs.BufferList = recvBufferList;
+                    recvArgs.Completed += (_,__) => tcs.SetResult(true);
+
+                    for (int i = 1; i <= 10; i++)
+                    {
+                        tcs = new TaskCompletionSource<bool>();
+
+                        sendBuffer[0] = (byte)i;
+                        if (server.SendAsync(sendArgs))
+                        {
+                            await tcs.Task;
+                        }
+
+                        recvBuffer[0] = 0;
+                        tcs = new TaskCompletionSource<bool>();
+                        if (client.ReceiveAsync(recvArgs))
+                        {
+                            await tcs.Task;
+                        }
+
+                        Assert.Equal(1, recvArgs.BytesTransferred);
+                        Assert.Equal(i, recvBuffer[0]);
+
+                        // Mutate the send/recv BufferLists
+                        // This should not affect Send or Receive behavior, since the buffer list is cached
+                        // at the time it is set.
+                        sendBufferList[0] = new ArraySegment<byte>(sendBuffer, i, 1);
+                        recvBufferList[0] = new ArraySegment<byte>(recvBuffer, i, 1);
```

Comment:
Rather than using an array segment of array segments, if you used a list of segments, you could also add/insert segments here to show there's no impact from that either.